### PR TITLE
LoadEventNexus MPI support

### LINKED ISSUE #20555: Basic MPI support for LoadEventNexus
Follow-up to preparations in #20500. Add actual MPI support for `LoadEventNexus`. 

Part of #20344.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Adds MPI support for `LoadEventNexus`. The implementation is not particularly efficient and not meant for production. An actual parallel loader will be added later.

Depends on https://github.com/mantidproject/mantid/pull/20511 (merged).

Note that `LoadInstrument` is now working around #9365: `LoadParameterFile` causes segfaults in MPI related unit tests since it is modifying the base instrument and is thus not thread safe. This should not be an issue in an actuall MPI run.

**To test:**

Code review.

Fixes #20555.

No release notes, MPI is not officially supported yet.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by OwenArnold — COMMENTED** [review]
I'll re-review when #20511 is merged as tricky to distinguish changes at this point.

**Comment by SimonHeybrock** [comment]
@OwenArnold Thanks, you know that you can select a sub range commits on the top left in the github diff tab?

File: Framework/API/inc/MantidAPI/TriviallyParallelAlgorithm.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_API_DLL TriviallyParallelAlgorithm : public Algorithm {
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply Framework/API/inc/MantidAPI/TriviallyParallelAlgorithm.h]: No, I think last time I said I didn't do it since there was only a single ported algorithm? Now there are 4 so it makes sense, I think, unless you have an argument against it?

**Comment by OwenArnold** [comment]
@SimonHeybrock No I didn't know that! but now that I've done that, I'll continue.

File: Framework/API/inc/MantidAPI/TriviallyParallelAlgorithm.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (OwenArnold) [reply Framework/API/inc/MantidAPI/TriviallyParallelAlgorithm.h]: The base class makes sense, definitely keep it.

File: Framework/DataHandling/inc/MantidDataHandling/LoadInstrument.h [hunk 2]
```diff
@@ -75,10 +69,8 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 File change history is stored at: <https://github.com/mantidproject/mantid>
 */
-class DLLExport LoadInstrument : public API::Algorithm {
+class DLLExport LoadInstrument : public API::TriviallyParallelAlgorithm {
```
Comment (OwenArnold) [inline Framework/DataHandling/inc/MantidDataHandling/LoadInstrument.h]: Why is this change necessary for LoadInstrument?

Comment (SimonHeybrock) [reply Framework/DataHandling/inc/MantidDataHandling/LoadInstrument.h]: It is used by `LoadEventNexus` to load the instrument and must thus support MPI to have MPI support in `LoadEventNexus`. The same goes for the other modified algorithms.

Comment (OwenArnold) [reply Framework/DataHandling/inc/MantidDataHandling/LoadInstrument.h]: Yes. Since the instrument itself will not be distributed, we must ensure that it is fully loaded on all ranks that need it. I think the `TriviallyParallelAlgorithm` will work correctly in this case. No other changes needed.

File: Framework/DataHandling/src/LoadInstrument.cpp [hunk 3]
```diff
@@ -192,16 +192,17 @@ void LoadInstrument::exec() {
...
 
-  // check if default parameter file is also present, unless loading from
-  if (!m_filename.empty())
-    runLoadParameterFile();
+    // LoadParameterFile modifies the base instrument stored in the IDS so this
+    // must also be protected by the lock until LoadParameterFile is fixed.
+    // check if default parameter file is also present, unless loading from
+    if (!m_filename.empty())
+      runLoadParameterFile();
+  }
```
Comment (OwenArnold) [inline Framework/DataHandling/src/LoadInstrument.cpp]: 👍

**Comment by OwenArnold** [comment]
Now that threading issue is solved. Looks good to me :shipit:

**Review by OwenArnold — APPROVED** [review]

File: Framework/API/inc/MantidAPI/TriviallyParallelAlgorithm.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/API/inc/MantidAPI/TriviallyParallelAlgorithm.h]: I like it :+1:
